pkg/reconciler/configrevision_reconciler.go

```diff
@@ -0,0 +1,466 @@
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/telekom/das-schiff-network-operator/api/v1alpha1"
+	"github.com/telekom/das-schiff-network-operator/pkg/debounce"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	StatusInvalid      = "invalid"
+	StatusProvisioning = "provisioning"
+	StatusProvisioned  = "provisioned"
+
+	controlPlaneLabel = "node-role.kubernetes.io/control-plane"
+	numOfRefs         = 2
+	configTimeout     = time.Minute * 2
+)
+
+// ConfigRevisionReconciler is responsible for creating NodeConfig objects.
+type ConfigRevisionReconciler struct {
+	logger      logr.Logger
+	debouncer   *debounce.Debouncer
+	client      client.Client
+	timeout     time.Duration
+	scheme      *runtime.Scheme
+	maxUpdating int
+}
+
+// Reconcile starts reconciliation.
+func (crr *ConfigRevisionReconciler) Reconcile(ctx context.Context) {
+	crr.debouncer.Debounce(ctx)
+}
+
+// // NewNodeConfigReconciler creates new reconciler that creates NodeConfig objects.
+func NewNodeConfigReconciler(clusterClient client.Client, logger logr.Logger, timeout time.Duration, s *runtime.Scheme, maxUpdating int) (*ConfigRevisionReconciler, error) {
+	reconciler := &ConfigRevisionReconciler{
+		logger:      logger,
+		timeout:     timeout,
+		client:      clusterClient,
+		scheme:      s,
+		maxUpdating: maxUpdating,
+	}
+
+	reconciler.debouncer = debounce.NewDebouncer(reconciler.reconcileDebounced, defaultDebounceTime, logger)
+
+	return reconciler, nil
+}
+
+func (crr *ConfigRevisionReconciler) reconcileDebounced(ctx context.Context) error {
+	revisions, err := listRevisions(ctx, crr.client)
+	if err != nil {
+		return fmt.Errorf("error listing revisions: %w", err)
+	}
+
+	nodes, err := listNodes(ctx, crr.client)
+	if err != nil {
+		return fmt.Errorf("error listing nodes: %w", err)
+	}
+
+	nodeConfigs, err := crr.listConfigs(ctx)
+	if err != nil {
+		return fmt.Errorf("error listing configs: %w", err)
+	}
+
+	totalNodes := len(nodes)
+	for i := range revisions.Items {
+		if err := crr.processConfigsForRevision(ctx, nodeConfigs.Items, &revisions.Items[i], totalNodes); err != nil {
+			return fmt.Errorf("failed to process configs for revision %s: %w", revisions.Items[i].Name, err)
+		}
+	}
+
+	revisionToDeploy := getFirstValidRevision(revisions.Items)
+
+	// there is nothing to deploy - skip
+	if revisionToDeploy == nil {
+		return nil
+	}
+
+	nodesToDeploy := getOutdatedNodes(nodes, nodeConfigs.Items, revisionToDeploy)
+
+	if err := crr.updateQueueCounters(ctx, revisions.Items, revisionToDeploy, len(nodesToDeploy)); err != nil {
```

Comment:
Have you tested whether this works? Considering you're updating status twice this might cause issues due to writing with an outdated generation counter failing on the first reconciliation whenever counters changed.

Reply:
I've tested it and didn't see any issues. However I've changed that bit so only one status update is performed now.